Parser/Client/Browser.php

```diff
@@ -664,14 +665,21 @@ public function parse(): ?array
             if ($name === $browserFromUserAgent['name']) {
                 $engine        = $browserFromUserAgent['engine'] ?? '';
                 $engineVersion = $browserFromUserAgent['engine_version'] ?? '';
+            }
 
-                // In case the user agent reports a more detailed version, we try to use this instead
-                if (!empty($browserFromUserAgent['version'])
-                    && 0 === \strpos($browserFromUserAgent['version'], $version)
-                    && \version_compare($version, $browserFromUserAgent['version'], '<')
-                ) {
-                    $version = $browserFromUserAgent['version'];
-                }
+            // In case the user agent reports a more detailed version, we try to use this instead
+            if (!empty($browserFromUserAgent['version'])
+                && 0 === \strpos($browserFromUserAgent['version'], $version)
+                && \version_compare($version, $browserFromUserAgent['version'], '<')
+            ) {
+                $version = $browserFromUserAgent['version'];
+            }
```

Comment:
This shouldn't be moved outside of the if condition. Let's assume the client hints report a `Firefox 16`, but the useragent is spoofed by some tool and contains `Safari 20.18`. We would then end up using `Firefox 20.18`, which is wrong.

Reply:
And, how can we detect the full version then? Some times, client hints don't provide `Sec-CH-UA-Full-Version`

Reply:
In most cases the browser reported in client hints and in user agent should match. In that case the version from user agent will be used. If not I'm not sure if we should simply use the version from useragent.

Reply:
it turns out to be interesting here, opera prints all versions in clienthints

04/04/2022
last version Opera Desktop Linux 85
last version Opera Mobile 68.2

Opera on Windows | Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/100.0.4896.60 Safari/537.36 OPR/85.0.4341.18Mozilla/5.0 (Windows NT 10.0; WOW64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/100.0.4896.60 Safari/537.36 OPR/85.0.4341.18
-- | --
Opera on Macos | Mozilla/5.0 (Macintosh; Intel Mac OS X 12_3_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/100.0.4896.60 Safari/537.36 OPR/85.0.4341.18
Opera on Linux | Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/100.0.4896.60 Safari/537.36 OPR/85.0.4341.18
Opera on Android | Mozilla/5.0 (Linux; Android 10; VOG-L29) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/100.0.4896.58 Mobile Safari/537.36 OPR/63.3.3216.58675
-- | --
Opera on Android | Mozilla/5.0 (Linux; Android 10; SM-G970F) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/100.0.4896.58 Mobile Safari/537.36 OPR/63.3.3216.58675Mozilla/5.0 (Linux; Android 10; SM-N975F) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/100.0.4896.58 Mobile Safari/537.36 OPR/63.3.3216.58675
if you look at this list, they just printed versions for all OS
```
Sec-CH-UA: '"Chromium";v="98", "Opera";v="82", " Not A;Brand";v="99", "OperaMobile";v="68"'
Sec-CH-UA-Platform: 'Android'
```

the current tests I am satisfied with the result.

But in the future it will be necessary to come up with something else when useragent is gone.
as an option:
if Mobile OS + opera family geе clienthint `OperaMobile=68`
if Desktop OS + opera famil get `Opera=82`